### PR TITLE
(GlobalScale) lock federation to internal

### PR DESCRIPTION
**Pull request by ArtificialOwl**
This is a long awaited feature: having `gs.federation = 'internal'` working.
In a **GlobalScale** setup, in order to have its users being able to share files to users from another instance of Nextcloud, admin must enable the global federation shares. Meaning that it is not possible to enable the federation share only for _local_ instances defined within the lookup-server. 
As a side note, the value is currently 'internal' by default, meaning that if the admin does not set the `gs.federation` to `'external'` (and open the federated shares to the outside), shares between local instances of GS are not available. This PR respects the default value.

This PR will allow to manage those entries in `config/config.php`:

`gs.federation = 'internal'`  
_to limit all shares only to local instances of the GS._

`gs.federation.incoming = 'internal'`  
_to limit incoming shares only to local instances of the GS_

`gs.federation.outgoing = 'internal'`  
_to limit incoming shares only to local instances of the GS_

### How is it working:

- if the config is set to internal, federated shares a allowed without checking the settings related to federated shares.

- When creating a federated shares, if the outgoing shares is limited to internal, we check that the address of the remote recipient is considered as local before initiating the generation of the federated share. 

- To avoid some spoofing from an external request, a password is added to the share during the exchange, but not stored. The password is based on the token of the share and the jwt key from the `config/config.php`. The password is added regardless of the limitation to internal outgoing shares, as long as the address of the recipient is known by the lookup-server.

- When receiving a federated share, if the incoming shares is limited to internal, we compare the password of the share.

This way, every instances of the GlobalScale setup can have their own setup regarding internal/external federated shares.

### WIP

Tests are not yet implemented, this edit affect so many files and test might ask for a lot of work, please review and confirm the concept first.

### dependencies

To retrieve the local instances of a LUS: https://github.com/nextcloud/lookup-server/pull/43

### CODE REVIEW

File: lib/private/GlobalScale/Config.php [hunk 1]
```diff
@@ -66,7 +82,127 @@ public function onlyInternalFederation() {
...
+		$lookup = $this->config->getSystemValue('lookup_server', '');
+
+		$request = new Request('/instances', Request::TYPE_GET);
+		$request->setAddressFromUrl($lookup);
+
+		try {
+			return $instances = $this->retrieveJson($request);
+		} catch (RequestContentException | RequestNetworkException | RequestResultSizeException | RequestServerException | RequestResultNotJsonException $e) {
+			\OC::$server->getLogger()
+						->log(2, 'Issue while retrieving instances from lookup: ' . $e->getMessage());
```
Comment (kesselb) [inline lib/private/GlobalScale/Config.php]:
```suggestion
						->warning('Issue while retrieving instances from lookup: ' . $e->getMessage());
```

Or if you want to stick with `log`: 
https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/public/ILogger.php#L54

File: apps/federatedfilesharing/lib/FederatedShareProvider.php [hunk 2]
```diff
@@ -977,13 +995,16 @@ public function userDeletedFromGroup($uid, $gid) {
...
 	 *
+	 * @param string $remote
+	 *
 	 * @return bool
 	 */
 	public function isOutgoingServer2serverShareEnabled() {
-		if ($this->gsConfig->onlyInternalFederation()) {
-			return false;
+		if ($this->gsConfig->onlyInternalFederation(IGlobalScaleConfig::OUTGOING)) {
+			return true;
```
Comment (juliushaertl) [inline apps/federatedfilesharing/lib/FederatedShareProvider.php]: If nothing is configured the default of `'internal'` will now always be true which means that the general files_sharing settings will be ignored. We should only take the internal setting into consideration if gs is enabled. Same for the following checks in this file.

Comment (juliushaertl) [reply apps/federatedfilesharing/lib/FederatedShareProvider.php]: Ah no, missed the check in https://github.com/nextcloud/server/pull/19391/files#diff-492646fe4f5be48135e151ebe79104f8R77, but if only internal is allowed it should not return true here, shound't it?

Comment (daita) [reply apps/federatedfilesharing/lib/FederatedShareProvider.php]: your link is too vague, which part should not return true ?

Comment (juliushaertl) [reply apps/federatedfilesharing/lib/FederatedShareProvider.php]: Nevermind, I also failed to get that isOutgoingServer2serverShareEnabled should still return true as the actual external federation is checked with allowedOutgoingFederation later.

**Comment by rullzer** [comment]
I'm going to close this due to lack of activity.
Feel free to reopen if anybody wants to continue.

**Comment by daita** [comment]
Well, it would have been nice to have at least a review on the way to implement the solution

**Pull request closed without merging**